in_process_exporter_metrics: Add documentation for process_exporter_metrics (#1198)
fluent/fluent-bit-docs: pipeline/inputs/process-exporter-metrics.md

---
description: >-
  A plugin based on Process Exporter to collect process level of metrics of system
  metrics
---

# Process Exporter Metrics

[Prometheus Node Exporter](https://github.com/prometheus/node_exporter) is a popular way to collect system level metrics from operating systems, such as CPU / Disk / Network / Process statistics.
Fluent Bit 2.2 onwards includes a process exporter plugin that builds off the Prometheus design to collect process level metrics without having to manage two separate processes or agents.

The Process Exporter Metrics plugin implements collecting of the various metrics available from [the 3rd party implementation of Prometheus Process Exporter](https://github.com/ncabatoff/process-exporter) and these will be expanded over time as needed.

**Important note:** All metrics including those collected with this plugin flow through a separate pipeline from logs and current filters do not operate on top of metrics.

> This plugin is only supported on Linux based operating systems due to heavily relying on proc filesystem.

Reviewer: ```suggestion
This plugin is only supported on Linux based operating systems as it uses the `proc` filesystem to access the relevant metrics.
```

> macOS does not have procfs. Therefore, this plugin won't work for it.

Reviewer: ```suggestion
macOS does not have the `proc` filesustem so this plugin will not work for it.
```

# them through a Prometheus HTTP end-point.
#
# After starting the service try it with:
#
# $ curl http://127.0.0.1:2021/metrics
#
[SERVICE]
    flush           1
    log_level       info

[INPUT]
    name            process_exporter_metrics
    tag             process_metrics
    scrape_interval 2

[OUTPUT]
    name            prometheus_exporter
    match           process_metrics
    host            0.0.0.0
    port            2021
```

You can see the metrics by using _curl:_

```bash
curl http://127.0.0.1:2021/metrics
```

### Container to Collect Host Metrics

When deploying Fluent Bit in a container you will need to specify additional settings to ensure that Fluent Bit has access to the process details.
The following `docker` command deploys Fluent Bit with a specific mount path for `procfs` and settings enabled to ensure that Fluent Bit can collect from the host. 
These are then exposed over port 2021.

```
docker run -ti -v /proc:/host/proc:ro \
               -p 2021:2021        \
               fluent/fluent-bit:2.2 \
               /fluent-bit/bin/fluent-bit \
                         -i process_exporter_metrics -p path.procfs=/host/proc  \

>                          -o prometheus_exporter -p "add_label=host $HOSTNAME" \

Reviewer: we need to specify the `HOSTNAME` variable btw (`-e HOSTNAME="$HOSTNAME"`) or just remove it might be easier

Author: Sure. I removed it from the doc.

---
description: >-
  A plugin based on Process Exporter to collect process level of metrics of system
  metrics
---

# Process Exporter Metrics

> [Prometheus Node Exporter](https://github.com/prometheus/node_exporter) is a popular way to collect system level metrics from operating systems, such as CPU / Disk / Network / Process statistics. Fluent Bit 2.2 includes process exporter metrics plugin that builds off the Prometheus design to collect process level metrics without having to manage two separate processes or agents.

Reviewer: Let's split into a sentence per line for reviewing.
```suggestion
[Prometheus Node Exporter](https://github.com/prometheus/node_exporter) is a popular way to collect system level metrics from operating systems, such as CPU / Disk / Network / Process statistics.
Fluent Bit 2.2 onwards includes a process exporter plugin that builds off the Prometheus design to collect process level metrics without having to manage two separate processes or agents.
```

Explain what process level vs system level is.

> The initial release of Process Exporter Metrics refers a process group level collector to collect metrics available from [the 3rd party implementation of Prometheus Process Exporter](https://github.com/ncabatoff/process-exporter) and we plan to expand them over time if needed.

Reviewer: ```suggestion
The Process Exporter Metrics plugin implements collecting of the various metrics available from [the 3rd party implementation of Prometheus Process Exporter](https://github.com/ncabatoff/process-exporter) and these will be expanded over time as needed.
```

> **Important note:** Metrics collected with Process Exporter Metrics flow through a separate pipeline from logs and current filters do not operate on top of metrics.

Reviewer: This is true for all metrics plugins - we really should have a common page for it then just link out.

Reviewer: ```suggestion
**Important note:** All metrics including those collected with this plugin flow through a separate pipeline from logs and current filters do not operate on top of metrics.
```

> This plugin is only supported on Linux based operating systems.

Reviewer: What about macOS?

Author: macOS does not have procfs. So, we can support it only for Linux-based operating systems.

Reviewer: Can we add a comment to that effect? I would consider macOS to be in the "Linux-based" terminology so it is good to be clear it is not supported.

Author: macOS does not based on Linux-based technology. It *is* based on BSD 4.2 and Mach technology. filesystem is built on top of IOKit which is made by Apple actually....

Reviewer: Sure, I understand but I think it's easy to mistake it for "Linux-based" so be good to ensure we clarify this is not supported on macOS. That's my only point, clarify it is not supported because of the lack of `procfs`.

Author: Sounds reasonable. I added descriptions for procfs and won't supported it for macOS: https://github.com/fluent/fluent-bit-docs/pull/1198/commits/70ad3c2c905d1ce24578873ca5f09eb054e94266

## Configuration

| Key                       | Description                                                                            | Default   |
| ------------------------- | -------------------------------------------------------------------------------------- | --------- |

> | scrape_interval           | The rate at which metrics are collected from the host operating system                 | 5 seconds |

Reviewer: ```suggestion
| scrape_interval           | The rate at which metrics are collected.                 | 5 seconds |
```

> | path.procfs               | The mount point used to collect process information and metrics                        | /proc/    |

Reviewer: Any specification on the type of mount required, e.g. is read-only ok?

Reviewer: Probably want an example for containers here - we cannot mount `/proc` so have to customise.

Author: Right. I added a note for read-only attribute: https://github.com/fluent/fluent-bit-docs/pull/1198/commits/d8e073872e2bb35794ff7847ed2984ce905fabb8

| process\_include\_pattern | regex to determine which processes are included in the metrics produced by this plugin | It is applied for all process unless explicitly set. Default is `.+`. |

> | process\_exclude\_pattern | regex to determine which units are excluded in the metrics produced by this plugin     | It is not applied unless explicitly set. Default is `NULL`. |

Reviewer: I think these two descriptions slightly confuse things.

What is the regex applied to, the process name? What if you run multiple processes of the same name, how do you distinguish them?

The exclude pattern talks about `units` - are these systemd units? What about non-systemd deployments in that case (e.g. containers)?

## Getting Started

### Simple Configuration File

In the following configuration file, the input plugin _process\_exporter\_metrics collects _metrics every 2 seconds and exposes them through our [Prometheus Exporter](../outputs/prometheus-exporter.md) output plugin on HTTP/TCP port 2021.

```
# Process Exporter Metrics + Prometheus Exporter
# -------------------------------------------
# The following example collect host metrics on Linux and expose
# them through a Prometheus HTTP end-point.
#
# After starting the service try it with:
#
# $ curl http://127.0.0.1:2021/metrics
#
[SERVICE]
    flush           1
    log_level       info

[INPUT]
    name            process_exporter_metrics
    tag             process_metrics
    scrape_interval 2

[OUTPUT]
    name            prometheus_exporter
    match           process_metrics
    host            0.0.0.0
    port            2021
```

You can test the expose of the metrics by using _curl:_

```bash
curl http://127.0.0.1:2021/metrics
```

### Container to Collect Host Metrics

> When deploying Fluent Bit in a container you will need to specify additional settings to ensure that Fluent Bit has access to the host operating system. The following docker command deploys Fluent Bit with specific mount path for ptocfs and settings enabled to ensure that Fluent Bit can collect from the host. These are then exposed over port 2021.

Reviewer: Separate lines for review.
```suggestion
When deploying Fluent Bit in a container you will need to specify additional settings to ensure that Fluent Bit has access to the process details.
The following `docker` command deploys Fluent Bit with a specific mount path for `procfs` and settings enabled to ensure that Fluent Bit can collect from the host. 
These are then exposed over port 2021.
```

```

> docker run -ti -v /proc:/host/proc \

Author: We don't write any contents into procfs. So, we can use readonly option here.

>                -v /sys:/host/sys   \

Reviewer: Why do we need `/sys`? We have not mentioned it anywhere...

Author: Oh, I forgot to remove this...

               -p 2021:2021        \

>                fluent/fluent-bit:2.2.0 \

Reviewer: ```suggestion
               fluent/fluent-bit:2.2 \
```

The Process Exporter Metrics plugin implements collecting of the various metrics available from [the 3rd party implementation of Prometheus Process Exporter](https://github.com/ncabatoff/process-exporter) and these will be expanded over time as needed.

**Important note:** Metrics collected with Process Exporter Metrics flow through a separate pipeline from logs and current filters do not operate on top of metrics.

This plugin is only supported on Linux based operating systems due to heavily relying on proc filesystem.

macOS does not have procfs. Therefore, this plugin won't work for it.

## Configuration

| Key                       | Description                                                                            | Default   |
| ------------------------- | -------------------------------------------------------------------------------------- | --------- |
| scrape_interval           | The rate at which metrics are collected.                 | 5 seconds |
| path.procfs               | The mount point used to collect process information and metrics. Read-only is enough   | /proc/    |

> | process\_include\_pattern | regex to determine which names of processes are included in the metrics produced by this plugin | It is applied for all process unless explicitly set. Default is `.+`. |

Reviewer: Can we clarify the regex handling?

It would be useful to have examples for how to monitor specific processes.
Is it the binary name itself as reported by `ps` for example or something else (e.g. if symlinked with a different name?).
Can we/how do we filter processes with the same name but different args? 
I presume we cannot filter by pid either?

Author: 1. Yes. The binary names are handled as ps reported names which is limited 15 characters.
2. No. We cannot filtered by command line arguments like as prometheus process_exporter does.
3. Correct. We cannot filter by pid.

Reviewer: Thanks @cosmo0920.

Can we add some examples?

Does the max-length include the path to the binary? 15 seems super small...

Author: Yup. This seems to be super small and I thought it was wrong.... However, the specification is correct.

The limits is originated from: 
* https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L292
* https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L1073

And according to man 2 prctl:

> PR_SET_NAME (since Linux 2.6.9)
              Set  the  name  of the calling thread, using the value in the location pointed to by (char *) arg2.  The name can be up to 16 bytes long, including the terminating null byte.  (If the length of
              the string, including the terminating null byte, exceeds 16 bytes, the string is silently truncated.)  This is the same attribute that can be set via pthread_setname_np(3) and  retrieved  using
              pthread_getname_np(3).  The attribute is likewise accessible via /proc/self/task/[tid]/comm, where tid is the name of the calling thread.

Reviewer: Can we add these details in a sub-section maybe with some examples?

Author: Sure, I'll add them
